mlflow/pipelines/steps/split.py

```diff
@@ -191,7 +202,22 @@ def _run(self, output_directory):
                 importlib.import_module(post_split_module_name), post_split_fn_name
             )
             _logger.debug(f"Running {post_split_fn_name} on train, validation and test datasets.")
-            (train_df, validation_df, test_df) = post_split(train_df, validation_df, test_df)
+            (post_split_train_df, post_split_validation_df, post_split_test_df) = post_split(
+                train_df, validation_df, test_df
+            )
+            if (
+                is_subset(post_split_train_df, train_df)
+                and is_subset(post_split_validation_df, validation_df)
+                and is_subset(post_split_test_df, test_df)
+            ):
+                train_df = post_split_train_df
+                validation_df = post_split_validation_df
+                test_df = post_split_test_df
+            else:
+                raise MlflowException(
+                    f"Attempting to perform arbitrary ETL transform in {post_split_fn_name}. "
+                    f"Arbitrary transformations should go into the transform step."
+                )
```

Comment:
From ML-25309, I think we should review the interface in `steps/split.py` in the MLP eng sync before merging. It's possible that we want to introduce a more restrictive interface - e.g. a method that maps the DF to a boolean series. IMO, I think the current interface easier for users to reason about, but we should make sure to get stakeholder alignment.